### ISSUE TITLE
pre-commit hook not executed on cygwin

### ISSUE THREAD

**Issue by fholzer** [typicode/husky#675]
I'm running Windows 10, CYGWIN, terminal is mintty.
Node: v12.13.0
Git: 2.21.0
Bash: GNU bash, version 4.4.12(3)-release (x86_64-unknown-cygwin)

My `package.json`'s husky section:
````
  "husky": {
    "hooks": {
      "pre-commit": "exit 1"
    }
  }
````

`HUSKY_DEBUG=1 git commit` output:
```
husky:debug husky v4.2.3 - pre-commit
husky:debug Current working directory is /cygdrive/c/Users/fholzer/Documents/git/git-hook-test
husky > pre-commit (node v12.13.0)
husky > pre-commit hook failed (add --no-verify to bypass)
husky:debug npx --no-install husky-run exited with 0 exit code
husky:debug husky v4.2.3 - prepare-commit-msg
husky:debug Current working directory is /cygdrive/c/Users/fholzer/Documents/git/git-hook-test
husky:debug prepare-commit-msg config not found, skipping hook
husky:debug husky v4.2.3 - commit-msg
husky:debug Current working directory is /cygdrive/c/Users/fholzer/Documents/git/git-hook-test
husky:debug commit-msg config not found, skipping hook
husky:debug husky v4.2.3 - post-commit
husky:debug Current working directory is /cygdrive/c/Users/fholzer/Documents/git/git-hook-test
husky:debug post-commit config not found, skipping hook
[master 1bf28aa] this commit shouldn't work
 2 files changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 test/test3
 create mode 100644 test/test4
```

After adding some more debug output to `runner/bin.js` and `runner/index.js` (prefixed with `DEBUG`):
````
husky:debug husky v4.2.3 - pre-commit
husky:debug Current working directory is /cygdrive/c/Users/fholzer/Documents/git/git-hook-test
husky > pre-commit (node v12.13.0)
DEBUG runner/index.js: runCommand: >exit 1< exited with null code
DEBUG  {
  error: Error: spawnSync /bin/bash ENOENT
      at Object.spawnSync (internal/child_process.js:1041:20)
      at Object.spawnSync (child_process.js:607:24)
      at runCommand (C:\Users\fholzer\Documents\git\git-hook-test\node_modules\husky\lib\runner\index.js:33:35)
      at Object.run [as default] (C:\Users\fholzer\Documents\git\git-hook-test\node_modules\husky\lib\runner\index.js:69:16)
      at run (C:\Users\fholzer\Documents\git\git-hook-test\node_modules\husky\lib\runner\bin.js:11:40)
      at Object.<anonymous> (C:\Users\fholzer\Documents\git\git-hook-test\node_modules\husky\lib\runner\bin.js:20:1)
      at Module._compile (internal/modules/cjs/loader.js:956:30)
      at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
      at Module.load (internal/modules/cjs/loader.js:812:32)
      at Function.Module._load (internal/modules/cjs/loader.js:724:14) {
    errno: 'ENOENT',
    code: 'ENOENT',
    syscall: 'spawnSync /bin/bash',
    path: '/bin/bash',
    spawnargs: [ '-c', 'exit 1' ]
  },
  status: null,
  signal: null,
  output: null,
  pid: 0,
  stdout: null,
  stderr: null
}
husky > pre-commit hook failed (add --no-verify to bypass)
DEBUG runner/bin.js: index.js exited with 0 code
husky:debug npx --no-install husky-run exited with 0 exit code
husky:debug husky v4.2.3 - prepare-commit-msg
husky:debug Current working directory is /cygdrive/c/Users/fholzer/Documents/git/git-hook-test
husky:debug prepare-commit-msg config not found, skipping hook
husky:debug husky v4.2.3 - commit-msg
husky:debug Current working directory is /cygdrive/c/Users/fholzer/Documents/git/git-hook-test
husky:debug commit-msg config not found, skipping hook
husky:debug husky v4.2.3 - post-commit
husky:debug Current working directory is /cygdrive/c/Users/fholzer/Documents/git/git-hook-test
husky:debug post-commit config not found, skipping hook
[master 79b464f] this commit shouldn't work
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 test/test5
````

Based on this, it seems that my `SHELL` environment variable is `/bin/bash`, which makes [`spawnSync` line 43](https://github.com/typicode/husky/blob/v4.2.3/src/runner/index.ts#L43) fail with ENOENT.

When setting `SHELL=bash` (instead of full cygwin path to the bash binary) when running git commit it seems to resolve the issue. Though `SHELL` is set to `/bin/bash` by default. I'm wondering how hooks can work for anyone running cygwin... maybe I'm missing something...

Running `SHELL=bash HUSKY_DEBUG=1 git commit`:
````
husky:debug husky v4.2.3 - pre-commit
husky:debug Current working directory is /cygdrive/c/Users/fholzer/Documents/git/git-hook-test
husky > pre-commit (node v12.13.0)
DEBUG runner/index.js: runCommand: >exit 1< exited with 1 code
DEBUG runner/index.js: runCommand: spawnSync returned:  {
  status: 1,
  signal: null,
  output: [ null, null, null ],
  pid: 10472,
  stdout: null,
  stderr: null
}
husky > pre-commit hook failed (add --no-verify to bypass)
DEBUG runner/bin.js: index.js exited with 1 code
husky:debug npx --no-install husky-run exited with 1 exit code
````

So one important observation I made here is that if a hook script can't be executed and [`spawnSync`](https://github.com/typicode/husky/blob/v4.2.3/src/runner/index.ts#L43) fails, there is no checks for that and `status` will be null, so it'll enter the `if` block because `null !== 0` and print information about the hook failing, but the `runCommand` function will return 0 because `return status || 0` and the hook won't block the commit.

**Comment by fholzer** [comment]
Seems this was introduced in 0ffe785 (https://github.com/typicode/husky/commit/0ffe7853853dd7667ddc24a78469b92b5312aeb2#diff-b3100e79c8ec53bf766cead6682b3935R43)

**Comment by fholzer** [comment]
Also, hooks are not executed in powershell for the same reason. (Because husky is trying to run the hook command in a shell. In powershell there is no `SHELL` env var, and `sh` is not found)
````
DEBUG: external hook command: { cmd: 'sh', args: [ '-c', 'exit 1' ] }
DEBUG: {
  cpres: {
    status: 1,
    signal: null,
    output: [ null, null, null ],
    pid: 13412,
    stdout: null,
    stderr: null
  }
}
````

Not sure if/how you want to solve the cygwin problem (or maybe the `SHELL` "problem" in general to also resolve this for powershell). If you're open to me looking into this further, just let me know.

**Comment by fholzer** [comment]
I think using the `shell` option of [`child_process.spawnSync`](https://nodejs.org/api/child_process.html#child_process_child_process_spawnsync_command_args_options) seems like a good alternative to figuring out what shell to use in husky.

> `shell` &lt;boolean&gt; | &lt;string&gt; If `true`, runs `command` inside of a shell. Uses `'/bin/sh'` on Unix, and `process.env.ComSpec` on Windows. A different shell can be specified as a string. See Shell Requirements and Default Windows Shell. Default: `false` (no shell).

**Comment by xCykrix** [comment]
This could also be related to #607, as my original environment in that issue was with cygwin.

My original conclusion was, possibly, that Cygwin uses virtual drive mounting (`/cygdrive/c/`) and it has been unable to resolve paths it needs. Executing the same environment outside of cygwin acts as it should.

**Comment by typicode** [comment]
I just published a new version (`4.2.4`), could you retry?

**Comment by fholzer** [comment]
Version 4.2.4 fixes the issue for me. Thank you!
